### PR TITLE
Extend header with amount of used data shares

### PR DESCRIPTION
**Pull request by Wondertan**
Closes: #249

### CODE REVIEW

File: proto/tendermint/types/types.proto [hunk 1]
```diff
@@ -68,17 +68,18 @@ message Header {
   // hashes of block data
   bytes last_commit_hash = 6;  // commit from validators from the last block
   bytes data_hash        = 7;  // transactions
+  int64 data_shares      = 8;
```
Comment (liamsi) [inline proto/tendermint/types/types.proto:71]: maybe add in a comment here with a (permanent) link to the LL spec.
Also, while it makes sense to keep related fields closer to each other, the comment `// hashes of block data` implies that this is a hash. I'd suggest adding a newline in between `data_hash` and `data_shares`.

Also, the naming is not clear. IMO, we should aim for parity with the spec where feasible. here `availableDataOriginalSharesUsed` is used in the spec. Either use that name or something that is clearly related. IMO, `num_original_data_shares` would also be clear enough (cc @adlerjohn).

**Review by liamsi — CHANGES_REQUESTED** [review]

File: proto/tendermint/types/types.proto (hunk 1 above)
Comment (adlerjohn) [reply proto/tendermint/types/types.proto:71]: I agree naming parity with the specs should be preferred.

Comment (Wondertan) [reply proto/tendermint/types/types.proto:71]: @liamsi, Should I then revisit all the protos to check their names and order?

Comment (liamsi) [reply proto/tendermint/types/types.proto:71]: We should eventually do this bur not necessarily in this PR. Complete parity of all the protos is currently not possible yet as the data structures still differ too much from the ones in the spec anyways. Here it is sufficient to have parity with the name and maybe the position in relation to the other ones in the spec.

File: types/block.go [hunk 2]
```diff
@@ -268,6 +268,7 @@ func (b *Block) fillDataAvailabilityHeader() {
 
 	// return the root hash of DA Header
 	b.DataHash = b.DataAvailabilityHeader.Hash()
+	b.NumOriginalDataShares = int64(dataSharesLen)
```
Comment (liamsi) [inline types/block.go:271]: I might be wrong but what is returned here:
https://github.com/lazyledger/lazyledger-core/blob/ece16550a1faa124450001b77eb340cd4424374a/types/block.go#L1398
isn't exactly what the spec specifies as 
https://github.com/lazyledger/lazyledger-core/blob/ece16550a1faa124450001b77eb340cd4424374a/types/block.go#L1390-L1398
will also include the tail padding shares.
So I think `ComputeShares` should return `curLen` instead: 
https://github.com/lazyledger/lazyledger-core/blob/ece16550a1faa124450001b77eb340cd4424374a/types/block.go#L1386
(cc @adlerjohn)

**Review by liamsi — APPROVED** [review]
LGMT 👍🏼

**Pull request merged**